partiql-value/src/boxed_variant.rs

```diff
@@ -0,0 +1,93 @@
+use dyn_clone::DynClone;
+use dyn_hash::DynHash;
+use partiql_common::pretty::PrettyDoc;
+use std::error::Error;
+
+use crate::datum::{Datum, DatumCategoryOwned, DatumCategoryRef, DatumLower};
+use crate::Value;
+use pretty::{DocAllocator, DocBuilder};
+use std::fmt::{Debug, Display};
+
+pub type BoxedVariantError = Box<dyn Error>;
+
+pub type BoxedVariantResult<T> = Result<T, BoxedVariantError>;
+pub type BoxedVariantValueIntoIterator = Box<dyn Iterator<Item = DynBoxedVariant>>;
+
+pub type BoxedVariantValueIter<'a> =
+    Box<dyn 'a + Iterator<Item = BoxedVariantResult<&'a DynBoxedVariant>>>;
+
+// dyn
+
+pub type DynBoxedVariantTypeTag = Box<dyn DynBoxedVariantType>;
+
+pub trait DynBoxedVariantType: Debug + DynClone {
+    fn construct(&self, bytes: Vec<u8>) -> BoxedVariantResult<Box<dyn BoxedVariant>>;
+    fn name(&self) -> &'static str;
+}
+
+dyn_clone::clone_trait_object!(DynBoxedVariantType);
+
+pub trait DynBoxedVariantTypeFactory {
+    fn to_dyn_type_tag(self) -> DynBoxedVariantTypeTag;
+}
+
+// typed
+
+pub type BoxedVariantTypeTag<D> = Box<dyn BoxedVariantType<Doc = D>>;
```

Comment:
lingering clippy warning

```
unnecessary associated type bound for dyn-incompatible associated type

warning: unnecessary associated type bound for dyn-incompatible associated type
  --> partiql-value/src/boxed_variant.rs:36:60
   |
36 | pub type BoxedVariantTypeTag<D> = Box<dyn BoxedVariantType<Doc = D>>;
   |                                                            ^^^^^^^ help: remove this bound
   |
   = note: this associated type has a `where Self: Sized` bound, and while the associated type can be specified, it cannot be used because trait objects are never `Sized`
   = note: `#[warn(unused_associated_type_bounds)]` on by default
```

feel free to mark as resolved if addressed in later PR

Reply:
Removed in #542 : https://github.com/partiql/partiql-lang-rust/pull/542/files#diff-922163275420073431bed7c2cd1da3bf82d387cb1cbb1022c0cf74f9b36d4a1fL21